DEVDOCS-3922 suggested updates to foundation guide (#1570)
bigcommerce/dev-docs: docs/api-docs/partner/subscriptions/foundation-guide.md

# Subscription Foundation

> Subscription Foundation is an open-source BigCommerce "subscribe and save" application framework. It provides a template to scaffold custom subscription billing and invoicing solutions for your business or client. Subscription Foundation provides a default integration with Stripe Billing, which significantly accelerates your development time. 

Reviewer: @stanzikratelbc @traci707 

1. Let's drum up a more compelling verb than `provides` for the intro paragraph.
2. `, which significantly accelerates` is going to feel mean to devs who are struggling to implement parts of this. I feel like the better  options are the modal `, which can significantly accelerate` or to imply causality with the infinitive `to accelerate...`.

> Subscription Foundation utilizes the [Channels toolkit](/api-docs/channels/guide/overview#channels-toolkit), which lets the custom subscription channel appear in the **Channel Manager** view of a store's control panel, alongside other sales channels.

Reviewer: I like the structural change, but `lets... appear` anthropomorphizes either the `utilizing` or the `Channels toolkit`. Can we make the antecedent clearer when we're claiming causality?

## Configure accounts

To configure your accounts you need to make sure the following requirements are met:

1. Because there is a [store email address constraint](/api-docs/apps/guide/developer-portal#store-email-address-constraint?source=subscription-foundation) on draft and private apps, you need to make sure that both your sandbox store and your Dev Portal account use the same email address. 
2. The sandbox store must support multi-storefront sales.
3. To develop and test apps, you need a BigCommerce sandbox store. If you don't have one, [Create a Sandbox Store](/api-docs/partner/getting-started/create-a-sandbox-store?source=subscription-foundation).
4. To register apps and create app API accounts, you need a BigCommerce Developer Portal account. If you don't have one, create a [Dev Portal account](https://devtools.bigcommerce.com/?source=subscription-foundation).

## Fork and install the source repository 

> As a first step you need to fork the source repository, clone it to your development, and install the dpackages for your app

Reviewer: I like having an intro sentence here, but words are missing & there's a typo. Please clarify.

> **To fork and install the source repository**

Reviewer: this is the name of the section (which is also an anchor link). Why do we need a second bold heading?

Author: I've used this style in previous jobs to introduce a procedure. So it's not a heading but an intro style. We could leave it unbold

1. Fork the [subscription-foundation repository (GitHub)](https://github.com/bigcommerce/subscription-foundation) to your GitHub account.
2. Clone your fork to your local development environment.
3. Navigate to your root directory of your cloned repository and install the default packages for your app by issuing the following command:

```shell title="Install packages"
npm install
```
## Create an HTTPS tunnel

> After you have successfully installed the packages for your app, need to serve your app over a fully-qualified publicly accessible URL to connect the app to your sandbox store. We recommend using [ngrok](https://ngrok.com/docs), a free tool that uses HTTP tunneling to securely expose `localhost` ports to the public internet. 

Reviewer: ```suggestion
After you successfully install the packages, prepare to connect the app to your sandbox store. Because you must serve apps over fully-qualified publicly accessible URLs, this guide uses a tool called [ngrok](https://ngrok.com/docs) to open HTTP tunnels that securely expose `localhost` ports to the internet. 
```

The URLs will be plural because apps must expose multiple endpoints, `auth` and `load` at a minimum.

We don't have to solve this today, but this `serving apps over` pattern is in a lot of places and I wish we had a better way with a different preposition, or a different subject.

Reviewer: also, past tense

> To get started with ngrok, read the [Create an HTTPS Tunnel](/api-docs/apps/tutorials/sample-app-nextjs/step-2-connect#create-an-https-tunnel) tutorial.

Reviewer: ```suggestion
To get started with ngrok, follow the [Create an HTTPS Tunnel](/api-docs/apps/tutorials/sample-app-nextjs/step-2-connect#create-an-https-tunnel) section of our more general sample app tutorial.
```

Author: let's avoid using "our". I think it would be good to say follow the.... section of the sample app tutorial.

> After you have successfully started ngrok, make sure to note your ngrok ID. You will use it later.

Reviewer: ```suggestion
After you successfully start ngrok, make sure to note your ngrok ID. You will use it later.
```

Reviewer: present tense preferred.

## Create an app profile

> As a next step, you need to create a draft app to generate app API account credentials. 

Reviewer: ```suggestion
As a next step, you need to create an app profile to generate app API account credentials. 
```

As I've been revising Developer Portal docs recently, I've realized there's a distinction between developing an app and registering its profile to communicate with BigCommerce about it. So I want to shift language to talk about creating or editing the `profile` rather than the app itself.

Reviewer: I could also go `draft app profile`.

To develop and test the app, you need to install it on your [sandbox store](#configure-accounts). 

> **To create a draft app**

Reviewer: again, what purpose does this second bold heading serve?

1. Sign in to the Dev Portal. Then follow the directions to [create an app](/api-docs/apps/guide/developer-portal#create-an-app?source=subscription-foundation). Enter the following values on the **Technical** tab.

> 2. In the **Callback URLs** section, supply the following URLs, replacing `{ngrok_id}` with your ngrok ID which you noted in an earlier step.

Reviewer: ```suggestion
2. In the **Callback URLs** section, supply the following URLs. Replace `{ngrok_id}` with the ngrok ID you noted in an earlier step.
```
General grammar soapbox items: 
* This kind of `which` construction is common, but isn't grammatically correct. 
* It's clearer and stronger to avoid linking phrases together with present progressives. We want to keep verbs as close to simple present as the grammatical situation allows.

The ngrok ID can (and will) change, so best to avoid possessive articles lest devs get attached and feel slighted when they have to edit the app profile later.

| Sites & Routes | read-only | `store_sites_read_only` |
| Storefront API Tokens | generate tokens | `store_storefront_api` |

4. On the **App Supported Features** tab, select the option that indicates this is a multi-storefront enabled app.

5. Click **Update & Close**, then click **Confirm Update** to acknowledge the OAuth scopes you configured.

> Make sure to remain signed into the Dev Portal

Reviewer: ```suggestion
Make sure to remain signed in to the Dev Portal.
```

## Configure Stripe

> Subscription Foundation uses Stripe Connect so that your app can use one signle connection to submit payments to multiple previously configured Stripe merchant accounts. See [Stripe Billing](https://support.bigcommerce.com/s/article/Connecting-Stripe-Payment-Gateway?language=en_US#billing) for merchant configuration instructions.

Reviewer: ```suggestion
Subscription Foundation uses Stripe Connect so that your app can use one single connection to submit payments to the previously configured Stripe accounts of multiple BigCommerce merchants. See [Stripe Billing](https://support.bigcommerce.com/s/article/Connecting-Stripe-Payment-Gateway?language=en_US#billing) for merchant configuration instructions.
```

I want to be clear that the `merchants` are the same BigCommerce merchants we reference elsewhere, and that there isn't such a thing as a `Stripe merchant`.

What's the reason to ax the possessive? IMO, it's fewer words with clearer relationships, so I'm interested in the feeling or rule behind this pattern I'm seeing throughout.

The integration between Stripe and BigCommerce requires that the app uses two keys to issue charges: the **merchant-specific Stripe account's public key** and the **app-specific Stripe account's secret key**.

> Using two separate keys provides the following benefits:

Reviewer: ```suggestion
Using two separate keys provides the following benefits:
```
it's not the twoness of the keys that provide the benefits. see my previous comment.

* The merchant can change between Stripe accounts they use without breaking the integration.
* The merchant's own Stripe account stores the entire shopper payment history, including subscription and one-time purchases. 
* The usage of public and secret keys ensures that your configuration is PCI compliant.

> For development you will use two Stripe accounts:

Reviewer: simplify verb tense

Reviewer: also I don't love numbering these accounts unless we're going to say that the numbers are based on the order the developer creates the account.

### Create Stripe accounts

To get started, complete the following steps:

1. To create the Stripe account that simulates your merchant's pre-existing Stripe account, [sign up with Stripe](https://dashboard.stripe.com/register). Your app uses this account's client ID to process subscriber payments and manage subscriptions. Make sure to give it a merchant-specific name.

2. Sign in to the [Stripe Dashboard](https://dashboard.stripe.com/).

3. On the [Stripe Dashboard](https://dashboard.stripe.com/), create and configure a second app-specific Stripe account. The app uses this account's client secret to process subscription payments.

   a. On the top left corner of your Stripe Dashboard, click the name of your merchant-specific account. Select **+ New Account** from the dropdown menu.

   b. Give the new account a name that clearly identifies it as app-specific to differentiate it from the merchant-specific payment and subscription management account. This guide uses **BigCommerce App** as an example.
   
   ![BigCommerce App](https://storage.googleapis.com/bigcommerce-production-dev-center/images/BigCommerce-app-image.png "The Stripe dashboard's new account dropdown menu")
   

>    c. Under **Get started with Stripe**, copy the **Secret key** and make sure to store it in a safe place. In a later step, you will use the secret key to update the `STRIPE_SECRET_KEY` environment variable in the .env file.

Reviewer: ```suggestion
   c. Under **Get started with Stripe**, copy the **Secret key** and make sure to store it in a safe place. In a later step, you will use the secret key to update the `STRIPE_SECRET_KEY` environment variable in the `.env` file.
```

### Enable Stripe Connect for Platforms

**During this following steps, make sure that you are logged into your the app-specific account by ehcking the top left corner of your Stripe Dashboard.**

<!-- theme: info -->
> #### Test mode
> You can simulate transactions in test mode to confirm your integration works correctly. 

**To configure test mode**

1. On your Stripe Dashboard click **Connect** to enable [Stripe Connect for Platforms](https://dashboard.stripe.com/test/connect/accounts/overview). 

2. Click **Get started**.

3. Select **Platform or Marketplace**, then click **Continue**.

> 4. To configure the app-specific account's Stripe Connect settings, on your Stripe Dashboard go to [**Settings > Connect settings**](https://dashboard.stripe.com/test/settings/connect). 

Reviewer: ```suggestion
4. To configure the app-specific account's Stripe Connect settings, go to [**Settings > Connect settings**](https://dashboard.stripe.com/test/settings/connect) on your Stripe Dashboard. 
```

5. Under **Integration** do the following:
   
   a. Copy the **Test mode client ID** and make sure to store it at a safe place. You will use the client ID in a later step to update `NEXT_PUBLIC_STRIPE_CLIENT_ID` in the .env file.

   b. Set **OAuth settings** to **OAuth for Standard accounts**. 

>    c. Click **Redirects**, and then click **+ Add URI**. Add the callback URI for your app:

Reviewer: ```suggestion
   c. Click **Redirects**, then click **+ Add URI**. Add the following callback URI for your app:
```

   
   ```http title="Your app's Stripe callback URI"
   https://{ngrok_id}.ngrok.io/stripe/callback
   ```

## Configure the BigCommerce store

> After you have successfully created a test configuration, you need to configure your BigCommerce store in the BigCommerce store control panel.

Reviewer: ```suggestion
After you successfully create a test configuration, configure your BigCommerce sandbox store in the store control panel.
```

Reviewer: what does `test configuration` mean here? It's vague. Be specific re what we accomplished in the preceding section.

**To configure the store to make subscription charges**

> 1. In the BigCommerce store control panel, in the [Stripe settings section](https://login.bigcommerce.com/deep-links/settings/payment/stripev3), make sure that **Test Mode** is set to **Yes**.

Reviewer: ```suggestion
1. In the BigCommerce store control panel, navigate to the [Stripe settings section](https://login.bigcommerce.com/deep-links/settings/payment/stripev3) and  make sure that **Test Mode** is set to **Yes**.
```

   
   ![stripe-settings](https://storage.googleapis.com/bigcommerce-production-dev-center/images/stripe-settings.png)
   

> 2. To start a subscription, shoppers must check out using a stored credit card. To enable this functionality, go to [Stripe settings](https://login.bigcommerce.com/deep-links/settings/payment/stripev3), locate the **Stored Credit Cards** section, and toggle on **Enable stored credit cards with Stripe**. 

Reviewer: `stored card` is a specific term in our literature. Not all stored cards are credit cards; we have stored debit cards and PayPal accounts, for example. If we really want to ax `stored card`, we could say `stored payment instrument`. See the Payment Processing doc.

   
   ![stored-credit-cards](https://storage.googleapis.com/bigcommerce-production-dev-center/images/stored-credit-cards.png)
   
3. To verify that the stored credit card functionality is enabled, go to your BigCommerce store, put some merchandise in your card. Then sign in to the BigCommerce store as a shopper, or create a new account during checkout. During checkout, in the payments step, select **Save this card for later**. 

## Declare environment variables

> 

Reviewer: How about an intro sentence that orients the reader back to working on the command line?

> #### Database note
> This example uses SQLite as a data store. In production, we recommend using a database that has more robust concurrency support, such as PostgreSQL. For information on switching databases, see the [Replacing SQLite](#replacing-sqlite) section.

1. If you're using SQLite, which is the default data store for Subscription Foundation, skip to the next step. Otherwise, follow the instructions in [Replacing SQLite](#replacing-sqlite)

> 2. Run the pre-configured Prisma migration script to create the database tables and initial client as defined in `/prisma/migrations/*` by issuing the following command:

Reviewer: 1. issuing again
2. the troubleshoot note is, I think, to encourage goodwill and perseverance. migrations rarely work the first time and tend to require a bunch of side work and other terminal commands.

```shell title="Run Prisma migration"
npx prisma migrate dev
```
The following screenshot shows a successful migration.
![npx_prisma_migrate](https://storage.googleapis.com/bigcommerce-production-dev-center/images/npx_prisma_migrate.png)

> 3. Start the app server by issuing the following command:

Reviewer: ```suggestion
3. Start the app server with the following command:
```

ngrok http 3000
```

After the app server and ngrok are running, you can install the draft app on your sandbox store. For more information about installing and troubleshooting apps in development, read [Installing and launching an app](/api-docs/apps/tutorials/sample-app-nextjs/step-2-connect#install-and-launch-the-app).

## Replacing SQLite

To use an alternate SQL database, do the following:

1. Update the `/prisma/schema.prisma` file with a `provider` other than `SQLite`. For a list of options, read [Prisma's reference docs](https://prisma.io/docs/reference/tools-and-interfaces/prisma-schema/data-sources/).

2. In `/prisma/.env`, update the value of the `DATABASE_URL` variable to match your new database connection string.

> 3. Run the prisma migration script by issuing the following command:

Reviewer: ```suggestion
3. Run the prisma migration script with the following command:
```

```shell title="Run Prisma migration"
`npx prisma migrate dev`
```

> 4. To generate a fresh app client that uses the new database provider issue the following following command:

Reviewer: ```suggestion
4. To generate a fresh app client that uses the new database provider, run the following script:
```

npm run dev
```
7. Start ngrok by issuing the following command:

```shell title="Start ngrok"
ngrok http 3000
```

## Managing subscription products

**To add new subscription rules and edit existing ones**

1. In the store control panel, navigate to the **Channel Manager** menu in the store control panel and click **Stripe Subscriptions**.

2. Click the product you want to modify. For more information, see [Foundations for Stripe Billing](https://support.bigcommerce.com/s/article/Connecting-Stripe-Payment-Gateway?language=en_US#billing).

<!-- theme: info -->
> #### Subscription sales channel
> The app will create its own dedicated sales channel upon installation.

If you plan to use the API to add products to the subscription sales channel, learn more about [product channel assignments](/api-docs/multi-storefront/api-guide#products).

## Troubleshooting

### Issue: Environment variable not found when running 

This error can occur when the `/prisma/schema.prisma` file contains the incorrect provider, or the `.env` file contains an incorrect `DATABASE_URL`. Verify that the information in both files is correct.

### Issue: A "Not found" error during app installation

> If you don't enable the proper OAuth scopes, the `/api/auth` request might fail. Go to the [Dev Portal](https://devtools.bigcommerce.com) to check out the OAuth scopes that are enabled in the profile of your app. To learn more, read [Managing Apps in the Developer Portal](/api-docs/apps/guide/developer-portal#edit-technical-details). Make sure your scopes are equal to or exceed those listed in the [Creating an app profile](#create-an-app-profile) section.

Reviewer: ```suggestion
If you don't enable the proper OAuth scopes, the `/api/auth` request might fail. Go to the [Dev Portal](https://devtools.bigcommerce.com) to check out the OAuth scopes that are enabled in your app's profile. To learn more, read [Managing Apps in the Developer Portal](/api-docs/apps/guide/developer-portal#edit-technical-details). Make sure your scopes equal or exceed those listed in the [Creating an app profile](#create-an-app-profile) section.
```
